EMB-442: tutorial for Device Provisioning (#71)
AVSystem/developer-zone: docs/Coiote_DM_Device_Onboarding/Device_Provisioning.md

# Automated Provisioning for Nordic boards

## Introduction
With Factory provisioning for Nordic IoT devices, you can load on-device communication credentials and any cloud-related configuration at the factory level to automate secure device onboarding to Coiote IoT DM cloud.

Here’s a tutorial to get you started with device provisioning using a dedicated script to be found in the [Anjay Zephyr Client repository](https://github.com/AVSystem/Anjay-zephyr-client).

## Prerequisites
- A Nordic board connected to your computer. 

> - Installed [Go Programming language](https://go.dev/dl).

Reviewer: `export PATH=$PATH:home/<USER>/go/bin` might be needed, to be verified with some fresh system and installation step by step...

Reviewer: checked on a fresh system, copy-pasting commands install mcumgr in `/home/<USER>/go/bin` and does not suggest adding it to `PATH`

Author: Added note about this.

2. Get the Coiote DM Access Token

    The provisioning script can register your device to Coiote DM automatically. You might use this option for the sake of this tutorial, but this is an optional step.

    !!! note
        If you wish to skip device registration to Coiote DM, then call `ptool.py` without `-t` and `-S` options.

    First an access token needs to be generated.

    - Create `get_token.sh` file:
        ```
        #!/bin/bash

        SERVER="https://eu.iot.avsystem.cloud"

        echo "Enter your login credentials for $SERVER"
        read -p "Login: " USER
        read -p "Password: " -s PASS

        curl -X POST \
           -H "Content-Type:application/x-www-form-urlencoded" \
           --data-urlencode "grant_type=password" \
           --data-urlencode "username=$USER" \
           --data-urlencode "password=$PASS" \
           "$SERVER/api/auth/oauth_password"
        ```
    - If you're using Linux, run `chmod u+x get_token.sh` to give execute rights. Under Windows you can use the GUI to allow execution of this file.

>     - Run `./get_token.sh`. The script will ask you for your login and password for eu.iot.avsyste.cloud, please provide it.

Author: Fixed

>     If a JSON structure had been displayed containing `"access_token"` your ready to proceed. Copy your token.

Reviewer: ```suggestion
    If a JSON structure containing `"access_token"` appears, you're ready to proceed. Copy your token.
```

Author: Fixed

# Automated Provisioning for Nordic boards

## Introduction
The device before it will be deployed in the field needs to be flashed and configured. This process takes some time and effort.

This tutorial will show you how to provision Nordic boards using the provisioning script found [Anjay-zephyr-client repositorie] (https://github.com/AVSystem/Anjay-zephyr-client).

## Prerequisites
- Nordic board connected to your computer.

> - Installed [mcumgr command line tool](https://docs.zephyrproject.org/3.1.0/services/device_mgmt/mcumgr.html).

Author: This is true. Go is required but this is kinda described in installation instructions for mcumgr but OK. Let it be your way.

- Zephyr development environment set up.
- Registered account on [EU IOT CLOUD](https://eu.iot.avsystem.cloud).

## Step 1: Getting Coiote Access Token

> The provision script can register the provisioned device to Coiote DM automaticlly. We will use this option in this tutorial but this is a optional step and can be omited. First an access token needs to be generated. 

Author: I have added a note how to run ptool.py without the registration to Coiote

1. Create `get_token.sh` file:
    ```
    #!/bin/bash

    if [ -z "$1" ]
    then
        echo "Usage:"
        echo "$0 [USERNAME]"
        exit 1
    fi

    USER="$1"
    SERVER="https://eu.iot.avsystem.cloud"

>     read -p "Password for user $USER on $SERVER: " -s PASS

Author: Done.

    curl -X POST \
       -H "Content-Type:application/x-www-form-urlencoded" \
       --data-urlencode "grant_type=password" \
       --data-urlencode "username=$USER" \
       --data-urlencode "password=$PASS" \
       "$SERVER/api/auth/oauth_password" 
    

>     ```

Author: Would `anjay-zephyr-client`  in `tools/provisioning_tool` be ok?

>     ```

Author: After some thought I would rather not add this script to the repo. Currently we are showing the user what is inside this script (he/she needs to paste it to a file). If I will add it to a repo this will become a blackbox that is asking for your credentials.

> 2. Run `chmod u+x get_token.sh` to give execute rights

Author: Added, tell me if it's ok now.

3. Run `./get_token.sh [USERNAME]` where `[USERNAME]` is the user registered on eu.iot.avsyste.cloud. The script will ask you for your password, please provide it.

    If a JSON structure had been displayed containing `"access_token"` your ready to proceed. Copy your token, it's valid only for a short period of time. 

    For more informaton how to aquire the access token see [REST API authentication](https://eu.iot.avsystem.cloud/doc/user/REST_API/REST_API_Authentication/).

## Step 2: Prepare configuration
Before running the script some configuration should be set. Example configuration can be found in `Anjay-zephyr-client/tools/provisioning-tool/configs` directory.

1. Edit `endpoint_cfg` contains LwM2M objects setting that will be uploaded to the device. Set `RID.Security.Mode` to `0` for PreShared Key or `2` for authentication using certificates. If PreShared Key authentication was choosen modify `RID.Security.PKOrIdentity` and `RID.Security.SecretKey` as well.

> 2. Edit `cert_info.json`. This file contains information for generating a self signed certificate. This configuration is needed only if `RID.Security.Mode` is set to `2` and the user don't want to provide certificates generated ealier.

Author: I spited the tutorial into PSK and Certificate sections.

> 3. Edit `lwm2m_server.json` modify `domain` entry to reflect your domain in Coiote server. This file is needed if you wish the script to automaticlly add the new device to Coiote DM.

Author: Yes so that is why I stated to modify the `domain` entry. The user should know what domain is he in.

> 3. Edit `lwm2m_server.json` modify `domain` entry to reflect your domain in Coiote server. This file is needed if you wish the script to automaticlly add the new device to Coiote DM.

Author: Spitted the tutorial into two sections PSK and Cert

## Provisioning the device using certificates
Now we will show how to provision the device using certificates. This method is very similar to the provisioning the device with PSK and will require just a few additional steps.

!!!note
    You may need to remove the device from Coiote if you finished the steps in previous section and the device is already registered. Coiote will not allow registration of the device with the same name.

1. Prepare configuration

    Like in the PSK example we will modify the configuration found in `Anjay-zephyr-client/tools/provisioning-tool/configs` directory.

    - Edit `endpoint_cfg` and change `RID.Security.Mode` to `2` for authentication using certificates.

    - Edit `lwm2m_server.json` modify `domain` entry to reflect your domain in Coiote server.

    - Edit `cert_info.json`. This file contains information for generating a self signed certificate. This configuration is needed only if user don't want to provide certificates generated ealier.

    - Get the certificate for `eu.iot.avsystem.cloud`. Run:

        `openssl s_client -showcerts -dtls eu.iot.avsystem.cloud:5684 > server.pem` to download server certificate and then

> 

Reviewer 2: I will move words "and then" to another line

> 

Reviewer 2: > Copy your token, it's valid only for a short period of time. 

I can't comment this part but if it's important maybe we should show this as important part (another line or sth)?

Author: done

3. Edit `lwm2m_server.json` modify `domain` entry to reflect your domain in Coiote server. This file is needed if you wish the script to automatically add the new device to Coiote DM.

4. If in `endpoint_cfg` file you set `RID.Security.Mode` as `2` (certificate), you need to provide also a certificate for the server. For `eu.iot.avsystem.cloud` run:

    `openssl s_client -showcerts -dtls eu.iot.avsystem.cloud:5684 > server.pem` to download server certificate and then 

    `openssl  x509 -outform der -in server.pem -out server.der` to convert it to DER format. 

## Step 3: Run provisioning tool
After creating the correct configuration for provisioning make sure that west configuration is correct and the `manifest.path` is set to an absolute path.
The script should be executed from a directory from which the command `west build` will work. For the tutorial we will use the `Anjay-zephyr-client/demo`.
We will assume also that you have nrf9160dk board but other Nordic boards supported by `Anjay-zephyr-client` can also be used.
    ```
    cd Anjay-zephyr-client/demo

>     ./../tools/provisioning-tool/ptool.py -b nrf9160dk_nrf9160_ns -s <SERIAL> \

Reviewer 2: Still is plain text. You should check how it looks on page.

Author: Discussed outside of this review. The page displays OK.

        -c ../tools/provisioning-tool/configs/endpoint_cfg -t <TOKEN> \
        -S ../tools/provisioning-tool/configs/lwm2m_server.json \
        -C ../tools/provisioning-tools/configs/cert_info.json -p server.der
    ```

> !!!note

Reviewer 2: I think it will work with indentation, but check this on page. At this moment I can't see this as a note part

Author: Discussed outside of this review. The page displays OK.
